server/src/main/java/org/opensearch/index/fielddata/fieldcomparator/UnsignedLongValuesComparatorSource.java

```diff
@@ -0,0 +1,156 @@
+    }
+
+    private NumericDocValues getNumericDocValues(LeafReaderContext context, BigInteger missingValue) throws IOException {
+        final SortedNumericDocValues values = loadDocValues(context);
+        if (nested == null) {
+            return FieldData.replaceMissing(sortMode.select(values), missingValue);
+        }
+        final BitSet rootDocs = nested.rootDocs(context);
+        final DocIdSetIterator innerDocs = nested.innerDocs(context);
+        final int maxChildren = nested.getNestedSort() != null ? nested.getNestedSort().getMaxChildren() : Integer.MAX_VALUE;
+        return sortMode.select(values, missingValue.longValue(), rootDocs, innerDocs, context.reader().maxDoc(), maxChildren);
+    }
+
+    @Override
+    public Object missingObject(Object missingValue, boolean reversed) {
+        if (sortMissingFirst(missingValue) || sortMissingLast(missingValue)) {
+            final boolean min = sortMissingFirst(missingValue) ^ reversed;
+            return min ? Numbers.MIN_UNSIGNED_LONG_VALUE : Numbers.MAX_UNSIGNED_LONG_VALUE;
+        } else {
+            if (missingValue instanceof Number) {
+                return ((Number) missingValue);
+            } else {
+                return new BigInteger(missingValue.toString());
+            }
+        }
+    }
+
+    @Override
+    public FieldComparator<?> newComparator(String fieldname, int numHits, boolean enableSkipping, boolean reversed) {
+        assert indexFieldData == null || fieldname.equals(indexFieldData.getFieldName());
+
+        final BigInteger lMissingValue = (BigInteger) missingObject(missingValue, reversed);
+        return new UnsignedLongComparator(numHits, null, null, reversed, false) {
+            @Override
+            public LeafFieldComparator getLeafComparator(LeafReaderContext context) throws IOException {
+                return new UnsignedLongLeafComparator(context) {
+                    @Override
+                    protected NumericDocValues getNumericDocValues(LeafReaderContext context, String field) throws IOException {
+                        return UnsignedLongValuesComparatorSource.this.getNumericDocValues(context, lMissingValue);
+                    }
+                };
+            }
+        };
+    }
+
+    @Override
+    public BucketedSort newBucketedSort(
+        BigArrays bigArrays,
+        SortOrder sortOrder,
+        DocValueFormat format,
+        int bucketSize,
+        BucketedSort.ExtraData extra
+    ) {
+        return new BucketedSort.ForUnsignedLongs(bigArrays, sortOrder, format, bucketSize, extra) {
+            private final BigInteger lMissingValue = (BigInteger) missingObject(missingValue, sortOrder == SortOrder.DESC);
+
+            @Override
+            public Leaf forLeaf(LeafReaderContext ctx) throws IOException {
+                return new Leaf(ctx) {
+                    private final NumericDocValues docValues = getNumericDocValues(ctx, lMissingValue);
+                    private long docValue;
```

Comment:
this too has to be BigInteger right ?

Reply:
In general - yes, but creating BigInteger is very expensive (comparing to primitive long), in every place where it is not needed (fe comparison), the `long` is used and `Long.xxxUnsigned` methods